### ISSUE TITLE
Improvement: Update references to e.which and e.keyCode to e.key (e.which and e.keyCode are deprecated)

### ISSUE THREAD

**Issue by Christopher-Rains** [materializecss/materialize#37]
### DEPRECATION

Per MDN on `event.which` and `event.keyCode`:

> **Warning**: This attribute is deprecated; you should use [`KeyboardEvent.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) instead, if available.

There are mixed uses of the `event.which` and `event.keyCode` properties throughout the source in the following locations:

**autocomplete.js** - references `event.keyCode`

- Line: 220 - https://github.com/materializecss/materialize/blob/v1-dev/js/autocomplete.js#L220
- Line: 225 - https://github.com/materializecss/materialize/blob/v1-dev/js/autocomplete.js#L225
- Line: 237 - https://github.com/materializecss/materialize/blob/v1-dev/js/autocomplete.js#L237
- Line: 240 - https://github.com/materializecss/materialize/blob/v1-dev/js/autocomplete.js#L240
- Line: 244 - https://github.com/materializecss/materialize/blob/v1-dev/js/autocomplete.js#L244

**datepicker.js** - references `event.which`

- Line: 812 - https://github.com/materializecss/materialize/blob/v1-dev/js/datepicker.js#L812

**dropdown.js** - references `event.which`

- Line: 224 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L224
- Line: 258 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L258
- Line: 263 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L263
- Line: 265 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L265
- Line: 286 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L286
- Line: 301 - https://github.com/materializecss/materialize/blob/v1-dev/js/dropdown.js#L301

**forms.js** - references `event.which`

- Line: 233 - https://github.com/materializecss/materialize/blob/v1-dev/js/forms.js#L233

**global.js** - references `event.which`

- Line: 44 - https://github.com/materializecss/materialize/blob/v1-dev/js/global.js#L44
- Line: 50 - https://github.com/materializecss/materialize/blob/v1-dev/js/global.js#L50

**timepicker.js** - references `event.which`

- Line: 139 - https://github.com/materializecss/materialize/blob/v1-dev/js/timepicker.js#L139

### IMPROVEMENT

An improvement to future proof these handlers for when `event.which` or `event.keyCode` is removed would be to use the suggested `event.key` property when attempting to identify which key was pressed. 

**Caveat - IE 11 implements non-standard key identifiers**

`event.key` differs from `event.which` or `event.keyCode` in that it returns the value of the key as opposed to the key's underlying code. Prior to the spec being approved, IE 11 implemented this change with non-standard values for the <kbd>ESC</kbd>, <kbd>Arrow Up</kbd>, and <kbd>Arrow Down</kbd> keys. To accomodate for this, one option would be to add the non-standard identifiers to the `M.keys` object in _global.js_:

```javascript
/* Lines 29 - 35 */
M.keys = {
  TAB: "Tab",
  ENTER: "Enter",
  ESC: "Escape",
  ESC_IE11: "Esc",
  ARROW_UP: "ArrowUp",
  ARROW_UP_IE11: "Up",
  ARROW_DOWN: "ArrowDown",
  ARROW_DOWN_IE11: "Down"
};
```

This would require a slight modification when one of these keys is pressed. For example, in _autocomplete.js_ on line 240:

```javascript
if ((key === M.keys.ARROW_UP || key === M.keys.ARROW_UP_IE11) && this.activeIndex > 0) {
  this.activeIndex--;
}
```

---

I'm happy to submit a PR for the changes above, but I wanted to get this in front of the **@materializecss/members** to determine whether the above solution is the preferred method to resolve this (I realize the IE 11 hack is not the most elegant) and where this fits in with the priority of your new build. I don't think this is critical and could wait a version or two if preferred.

**Comment by DanielRuf** [comment]
That sounds good. A PR is very welcome.

**Comment by schorsch** [comment]
Btw. i tried to get rid of those in another piece of code just a week ago and after three hours left them in place. 

Why? 

* they are still supported across browsers 
* there seems to be no prio, browser-wise, to kick those
* the alternatives are very verbose and not aligned cross-browser

just my 2 cents as i stumbled upon this fork today. Great to see some movement and i commented bcs. i think there are other parts which require more attention/man-power.

**Comment by DanielRuf** [comment]
> you should use [`KeyboardEvent.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) instead, if available.

**Comment by mauromascarenhas** [comment]
Closing this issue, since it has already been fixed on fea501a 😄.
